burstphoto/align.swift

```diff
@@ -40,22 +41,454 @@ let mfl = device.makeDefaultLibrary()!
+    pad_align_x = (pad_align_x*Int(tile_factor) - texture_width_orig)/2
+    
+    var pad_align_y = Int(ceil(Float(texture_height_orig)/Float(tile_factor)))
+    pad_align_y = (pad_align_y*Int(tile_factor) - texture_height_orig)/2
+      
+    // set reference texture
+    let ref_texture = extend_texture(textures[ref_idx], pad_align_x, pad_align_x, pad_align_y, pad_align_y)
+    
+    // build reference pyramid
+    let ref_pyramid = build_pyramid(ref_texture, downscale_factor_array)
+    
+    // blur reference texure and estimate noise standard deviation
+    // -  the computation is done here to avoid repeating the same computation in 'robust_merge()'
+    var ref_texture_blurred = blur_mosaic_texture(textures[ref_idx], kernel_size, mosaic_pattern_width)
+    ref_texture_blurred = blur_mosaic_texture(ref_texture_blurred, kernel_size, mosaic_pattern_width)
+    let noise_sd = estimate_color_noise(textures[ref_idx], ref_texture_blurred, mosaic_pattern_width)
+
+    // iterate over comparison images
+    for comp_idx in 0..<textures.count {
+        
+        // add the reference texture to the output
+        if comp_idx == ref_idx {
+            add_texture(textures[comp_idx], final_texture, textures.count)
+            DispatchQueue.main.async { progress.int += Int(80000000/Double(textures.count)) }
+            continue
+        }
+            
+        // set comparison texture
+        let comp_texture = extend_texture(textures[comp_idx], pad_align_x, pad_align_x, pad_align_y, pad_align_y)
+        
+        // check that the comparison image has the same resolution as the reference image
+        if (ref_texture.width != comp_texture.width) || (ref_texture.height != comp_texture.height) {
+            throw AlignmentError.inconsistent_resolutions
+        }
+       
+        // align comparison texture
+        let aligned_texture = crop_texture(align_texture(ref_pyramid, comp_texture, downscale_factor_array, tile_size_array, search_dist_array), pad_align_x, pad_align_x, pad_align_y, pad_align_y)
+        
+        // robust-merge the texture
+        let merged_texture = robust_merge(textures[ref_idx], ref_texture_blurred, aligned_texture, kernel_size, robustness, noise_sd, mosaic_pattern_width)
+        
+        // add robust-merged texture to the output image
+        add_texture(merged_texture, final_texture, textures.count)
+        
+        // sync GUI progress
+        DispatchQueue.main.async { progress.int += Int(80000000/Double(textures.count)) }
+    }
+}
+
+
+// convenience function for the frequency-based merging approach
+func align_merge_frequency_domain(progress: ProcessingProgress, shift_left: Int, shift_right: Int, shift_top: Int, shift_bottom: Int, ref_idx: Int, mosaic_pattern_width: Int, search_distance: Int, tile_size: Int, tile_size_merge: Int, robustness_norm: Double, read_noise: Double, max_motion_norm: Double, ft_mode: String, textures: [MTLTexture], final_texture: MTLTexture) throws {
```

Comment:
```suggestion
func align_merge_frequency_domain(progress: ProcessingProgress, shift_left_not_right: Bool, shift_top_not_bottom: Bool, ref_idx: Int, mosaic_pattern_width: Int, search_distance: Int, tile_size: Int, tile_size_merge: Int, robustness_norm: Double, read_noise: Double, max_motion_norm: Double, ft_mode: String, textures: [MTLTexture], final_texture: MTLTexture) throws {

    let shift_left   = shift_left_not_right ? tile_size_merge : 0;
    let shift_right  = shift_left_not_right ? 0 : tile_size_merge;
    let shift_top    = shift_top_not_bottom ? tile_size_merge : 0;
    let shift_bottom = shift_top_not_bottom ? 0 : tile_size_merge;
```

I don't think this should function should take the shifts in the different directions as separates variables. As far as I understand only one of `shift_right` or `shift_left` should be non-zero at a time. Passing a bool about which one should be non-zero guarantees this.

Reply:
I agree that it may be misleading and can be improved in a future version of the code.